external-crates/move/crates/move-compiler/src/cfgir/absint.rs

```diff
@@ -79,87 +39,169 @@ pub trait TransferFunctions {
+        &mut interpreter,
+        &CFGWrapper(cfg),
+        &(),
+        (initial_state, None),
+    )
+    .unwrap();
+    collect_states_and_diagnostics(inv_map)
+}
+
+struct AbstractInterpreter<'a, TF: TransferFunctions> {
+    pub transfer_functions: &'a mut TF,
+}
+
+impl<TF: TransferFunctions> absint::AbstractInterpreter for AbstractInterpreter<'_, TF> {
+    type Error = ();
+    type BlockId = Label;
+
+    type State = (<TF as TransferFunctions>::State, Option<Rc<Diagnostics>>);
+
+    type InstructionIndex = usize;
+    type Instruction = Command;
+
+    fn start(&mut self) -> Result<(), Self::Error> {
+        Ok(())
     }
 
-    fn execute_block(
+    fn join(
         &mut self,
-        cfg: &dyn CFG,
-        pre_state: &Self::State,
-        block_lbl: Label,
-    ) -> (Self::State, Diagnostics) {
-        let mut state = pre_state.clone();
-        let mut diags = Diagnostics::new();
-        for (idx, cmd) in cfg.commands(block_lbl) {
-            diags.extend(self.execute(&mut state, block_lbl, idx, cmd));
+        (pre, _): &mut (TF::State, Option<Rc<Diagnostics>>),
+        (post, _): &(TF::State, Option<Rc<Diagnostics>>),
+    ) -> Result<JoinResult, Self::Error> {
+        Ok(pre.join(post))
+    }
+
+    fn visit_block_pre_execution(
+        &mut self,
+        _block_id: Self::BlockId,
+        invariant: &mut absint::BlockInvariant<(TF::State, Option<Rc<Diagnostics>>)>,
+    ) -> Result<(), Self::Error> {
+        // each block needs its own unique Diagnostics, so set to None to be then initialized
+        // in the `execute`
+        invariant.pre.1 = None;
+        Ok(())
+    }
+
+    fn visit_block_post_execution(
+        &mut self,
+        _block_id: Self::BlockId,
+        _invariant: &mut absint::BlockInvariant<Self::State>,
+    ) -> Result<(), Self::Error> {
+        Ok(())
+    }
+
+    fn visit_successor(&mut self, _block_id: Self::BlockId) -> Result<(), Self::Error> {
+        Ok(())
+    }
+
+    fn visit_back_edge(
+        &mut self,
+        _from: Self::BlockId,
+        _to: Self::BlockId,
+    ) -> Result<(), Self::Error> {
+        Ok(())
+    }
+
+    fn execute(
+        &mut self,
+        lbl: Self::BlockId,
+        (start, _stop): (Self::InstructionIndex, Self::InstructionIndex),
+        (state, diags): &mut (TF::State, Option<Rc<Diagnostics>>),
+        idx: Self::InstructionIndex,
+        command: &Self::Instruction,
+    ) -> Result<(), Self::Error> {
+        if idx == start {
+            *diags = Some(Rc::new(Diagnostics::new()));
         }
-        (state, diags)
+        let diags: &mut Diagnostics = Rc::get_mut(diags.as_mut().expect("Set as some"))
+            .expect("unique ownership while iterating through block");
+        diags.extend(self.transfer_functions.execute(state, lbl, idx, command));
+        Ok(())
+    }
+}
+
+struct CFGWrapper<'a, T: CFG>(&'a T);
+impl<T: CFG> move_abstract_interpreter::control_flow_graph::ControlFlowGraph for CFGWrapper<'_, T> {
```

Comment:
Would it be possible to use this trait to replace `CFG` in `cfg.rs` eventually?

Reply:
I tried for a bit to get it to work, but I feel like it causes more pain that good. Mostly in that some of the return types are _slightly_ different. And the separation let's us more easily change the interface for compiler-centric datatypes.